File: use-cases/src/main/java/com/octo/usecases/CityForecastInteractor.java
package com.octo.usecases;

import com.octo.usecases.utils.StringUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class CityForecastInteractor {
    private static final int MINIMUM_RELEVANT_HOUR = 8;
    private static final int MAXIMUM_RELEVANT_HOUR = 20;
    private static final int REQUIRED_FORECASTS = 2;

    private final CityForecastRepository repository;
    private final CityForecastPresenter presenter;

    public CityForecastInteractor(final CityForecastRepository repository,
                                  final CityForecastPresenter presenter) {
        this.repository = repository;
        this.presenter = presenter;
    }

    public void loadCityForecast(final String cityName) {
        if (StringUtils.isBlank(cityName)) {
            presenter.onEmptyInput();
            return;
        }
        handleValidCityName(cityName);
    }

    private void handleValidCityName(String cityName) {
        try {
            handleRetrievedForecast(repository.loadCityForecast(cityName));
        } catch (CityForecastRepository.GenericException e) {
            presenter.onGenericException();
        } catch (CityForecastRepository.UnknownCityException e) {
            presenter.onUnknownCity();
        }
    }

    private void handleRetrievedForecast(CityWeeklyForecast cityForecast) {
        final List<Forecast> relevantForecasts = getRelevantForecasts(cityForecast);
        if (relevantForecasts.size() < REQUIRED_FORECASTS) {
            presenter.onUnavailableForecasts();
            return;
        }
        sortForecastsByAscendingTemperature(relevantForecasts);
        presenter.onForecasts(cityForecast.getCityName(), worstForecast(relevantForecasts), bestForecast(relevantForecasts));
    }

    private List<Forecast> getRelevantForecasts(CityWeeklyForecast cityForecast) {
        final List<Forecast> relevantForecasts = new ArrayList<>();
        for (Forecast forecast : cityForecast.getForecasts()) {
            if (isRelevant(forecast)) {
                relevantForecasts.add(forecast);
            }
        }
        return relevantForecasts;
    }

    private void sortForecastsByAscendingTemperature(List<Forecast> relevantForecasts) {
        Collections.sort(relevantForecasts, new Comparator<Forecast>() {
            @Override
            public int compare(Forecast f1, Forecast f2) {
                return f1.getTemperature().compareTo(f2.getTemperature());
            }
        });
    }

    private Forecast bestForecast(List<Forecast> sortedForecasts) {
        return sortedForecasts.get(sortedForecasts.size() - 1);
    }

    private Forecast worstForecast(List<Forecast> sortedForecasts) {
        return sortedForecasts.get(0);
    }

    private boolean isRelevant(Forecast forecast) {
        final Calendar calendar = Calendar.getInstance();
        calendar.setTime(forecast.getDate());
        final int hours = calendar.get(Calendar.HOUR_OF_DAY);
        return hours >= MINIMUM_RELEVANT_HOUR && hours <= MAXIMUM_RELEVANT_HOUR;
    }
}

File: use-cases/src/main/java/com/octo/usecases/CityForecastInteractor.kt
package com.octo.usecases

import com.octo.presentation.CityForecastDisplayableBuilder
import com.octo.presentation.ForecastDisplayable
import com.octo.repository.network.CityWeeklyForecast
import com.octo.repository.network.Forecast
import com.octo.repository.network.WeatherNetworkRepository
import java.io.IOException
import java.util.*
import javax.inject.Inject
import kotlin.collections.ArrayList

class CityForecastInteractor @Inject constructor (
    private val repository: WeatherNetworkRepository,
    private val displayableBuilder: CityForecastDisplayableBuilder
) {

    companion object {
        private const val MINIMUM_RELEVANT_HOUR = 8
        private const val MAXIMUM_RELEVANT_HOUR = 20
        private const val REQUIRED_FORECASTS = 2
    }

    sealed class State {
        data class Success(val displayable: ForecastDisplayable) : State()
        data class Error(val message: String) : State()
    }

    fun getCityForecastState(cityName: String) = if (StringUtils.isBlank(cityName)) {
        State.Error("Error city")
    } else {
        getHandledValidCityNameState(cityName)
    }

    private fun getHandledValidCityNameState(cityName: String): State = try {
        repository.loadCityWeeklyForecast(town = cityName)?.let { cityWeeklyForecast ->
            getHandleRetrievedForecastState(cityWeeklyForecast)
        } ?: State.Error("Error city")
    } catch (e: IOException) {
        State.Error("Error city")
    }

    private fun getHandleRetrievedForecastState(cityForecast: CityWeeklyForecast): State {
        val relevantForecasts: List<Forecast> = getRelevantForecasts(cityForecast)
        return if (getRelevantForecasts(cityForecast).size < REQUIRED_FORECASTS) {
            State.Error("Error city")
        } else {
            sortForecastsByAscendingTemperature(relevantForecasts)().let { sortedList ->
                State.Success(
                    displayable = displayableBuilder.buildForecastsDisplayable(
                        cityForecast.cityName,
                        worstForecast(sortedList),
                        bestForecast(sortedList)
                    )
                )
            }
        }
    }

    private fun getRelevantForecasts(cityForecast: CityWeeklyForecast): List<Forecast> = cityForecast.forecasts.mapNotNull { forecast ->
        if (isRelevant(forecast)) {
            forecast
        } else {
            null
        }
    }

    private fun sortForecastsByAscendingTemperature(relevantForecasts: List<Forecast>): () -> List<Forecast> = {
        relevantForecasts.sortedBy { it.temperature }
    }

    private fun bestForecast(sortedForecasts: List<Forecast>) = sortedForecasts[sortedForecasts.size - 1]

    private fun worstForecast(sortedForecasts: List<Forecast>) = sortedForecasts[0]

    private fun isRelevant(forecast: Forecast): Boolean {
        return getHourFromDate(forecast.date) in MINIMUM_RELEVANT_HOUR..MAXIMUM_RELEVANT_HOUR
    }

    private fun getHourFromDate(date: Date): Int {
        val calendar: Calendar = Calendar.getInstance()
        calendar.time = date
        return calendar.get(Calendar.HOUR_OF_DAY)
    }
}